### PR TITLE
Some buildmode additions, fixes

### PR DESCRIPTION
**Pull request by kriskog**
## About The Pull Request
Adds an outfit mode. Lets you select an outfit and apply it to humans with a left click. Right click to strip current outfit instead. Combine with custom outfits for best effect.
outfits

Adds a delete mode. Allows for left click deletion of any atom, including mobs. Right click to mass delete (once again including mobs), requires debug and server permission.
Expands the forbidden vars list of DuplicateObject, makes buildmode and podlauncher copy mode compatible with humans. No longer copies and shares organs, bodyparts, huds, etc.
copy

## Why It's Good For The Game
Massively simplifies outfitting several mobs for quickdraw events, simplifies quick deletion of mobs and expands the functionality of copy mode.

## Changelog
:cl: Skoglol
admin: Added a few new buildmode modes. Outfit, to quickly apply or remove outfits. Delete, to quickly delete any atoms. Slightly overhauled the copy proc used here and in supply pods, now compatible with most mobs.
/:cl:

### CODE REVIEW

File: code/modules/buildmode/submodes/delete.dm [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	var/list/pa = params2list(params)
+	var/left_click = pa.Find("left")
+	var/right_click = pa.Find("right")
+
+	if(left_click)
+		qdel(object)
+
+	if(right_click)
+		if(check_rights(R_DEBUG|R_SERVER))	//Prevents buildmoded non-admins from breaking everything.
+			var/action_type = alert("Strict type ([object.type]) or type and all subtypes?",,"Strict type","Type and subtypes","Cancel")
```
Comment (kriskog) [inline code/modules/buildmode/submodes/delete.dm:20]: Largely copied and adjusted for my needs from the similar VV dropdown option available on objects. Could probably be converted to a global proc and also expanded to all atoms in VV dropdown.

File: code/modules/buildmode/submodes/outfit.dm [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	dressuptime = c.robust_dress_shop()
+
+/datum/buildmode_mode/outfit/handle_click(client/c, params, obj/object)
+	var/list/pa = params2list(params)
+	var/left_click = pa.Find("left")
+	var/right_click = pa.Find("right")
+
+	if(!ishuman(object))
+		return
+	var/mob/living/carbon/human/dollie = object
```
Comment (kriskog) [inline code/modules/buildmode/submodes/outfit.dm:30]: No bully

File: code/modules/holodeck/area_copy.dm [hunk 3]
```diff
@@ -52,6 +54,11 @@ GLOBAL_LIST_INIT(duplicate_forbidden_vars,list(
 				contained_atom.flags_1 |= HOLOGRAM_1
 			if(M.circuit)
 				M.circuit.flags_1 |= HOLOGRAM_1
+
+	if(ismob(O))	//Overlays are carried over despite disallowing them, if a fix is found remove this.
+		var/mob/M = O
+		M.cut_overlays()
+		M.regenerate_icons()
```
Comment (kriskog) [inline code/modules/holodeck/area_copy.dm:61]: Attempted to disallow overlays in duplicate_forbidden_vars, but it wasnt working and copied them over anyway. If it hadnt, would likely need a regenerate_icons regardless. Suggestions welcome.

**Comment by kriskog** [comment]
Need to check for and likely track down some runtimes here as well, particularly from the copy mode. Will do so tomorrow.

**Comment by kriskog** [comment]
There are still the odd runtime if you copy humans and they are interacted with in some ways, but nothing game breaking. Wiped out a pretty noisy simplemob one, those seem to not cause issues anymore.

**Comment by spookydonut** [comment]
Holy shit this is amazing, well done.

File: code/modules/buildmode/submodes/outfit.dm [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+			return
+
+		for (var/obj/item/I in dollie.get_equipped_items(TRUE))
+			qdel(I)
+		if(dressuptime != "Naked")
+			dollie.equipOutfit(dressuptime)
+
+	if(right_click)
+		for (var/obj/item/I in dollie.get_equipped_items(TRUE))
+			qdel(I)
```
Comment (psykzz) [inline code/modules/buildmode/submodes/outfit.dm:44]: How about adding a middle click of shift click to copy someones current outfit?

Comment (kriskog) [reply code/modules/buildmode/submodes/outfit.dm:44]: Easier said than done, especially when dealing with inventories. Use custom outfits (made in game or in a json) to achieve something like this.

**Review by Twaticus — APPROVED** [review]

File: code/modules/buildmode/submodes/outfit.dm [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+		return
+	var/mob/living/carbon/human/dollie = object
+
+	if(left_click)
+		if(isnull(dressuptime))
+			to_chat(c, "<span class='warning'>Pick an outfit first.</span>")
+			return
+
+		for (var/obj/item/I in dollie.get_equipped_items(TRUE))
+			qdel(I)
```
Comment (Rohesie) [inline code/modules/buildmode/submodes/outfit.dm:38]:
```suggestion
		for (var/item in dollie.get_equipped_items(TRUE))
			qdel(item)
```

File: code/modules/buildmode/submodes/outfit.dm [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+		if(dressuptime != "Naked")
+			dollie.equipOutfit(dressuptime)
+
+	if(right_click)
+		for (var/obj/item/I in dollie.get_equipped_items(TRUE))
+			qdel(I)
```
Comment (Rohesie) [inline code/modules/buildmode/submodes/outfit.dm:44]:
```suggestion
		for (var/item in dollie.get_equipped_items(TRUE))
			qdel(item)
```

File: code/modules/holodeck/area_copy.dm [hunk 7]
```diff
@@ -20,7 +23,7 @@ GLOBAL_LIST_INIT(duplicate_forbidden_vars,list(
 			if(islist(original.vars[V]))
 				var/list/L = original.vars[V]
 				O.vars[V] = L.Copy()
-			else if(istype(original.vars[V], /datum))
+			else if(istype(original.vars[V], /datum) || istype(original.vars[V], /mob))
```
Comment (Rohesie) [inline code/modules/holodeck/area_copy.dm:26]:
```suggestion
			else if(istype(original.vars[V], /datum) || ismob(original.vars[V]))
```

File: code/modules/buildmode/submodes/delete.dm [hunk 8]
```diff
@@ -0,0 +1,62 @@
+/datum/buildmode_mode/delete
+	key = "delete"
+
+/datum/buildmode_mode/delete/show_help(client/c)
+	to_chat(c, "<span class='notice'>***********************************************************</span>")
+	to_chat(c, "<span class='notice'>Left Mouse Button on anything to delete it. If you break it, you buy it.</span>")
+	to_chat(c, "<span class='notice'>Right Mouse Button on anything to delete everything of the type. Probably dont do this unless you know what you are doing.</span>")
+	to_chat(c, "<span class='notice'>***********************************************************</span>")
```
Comment (Rohesie) [inline code/modules/buildmode/submodes/delete.dm:8]:
```suggestion
	to_chat(c, "<span class='notice'>***********************************************************\n\
		Left Mouse Button on anything to delete it. If you break it, you buy it.\n\
		Right Mouse Button on anything to delete everything of the type. Probably dont do this unless you know what you are doing.\n\
		***********************************************************</span>")
```

Comment (Rohesie) [inline code/modules/buildmode/submodes/delete.dm:8]:
```suggestion
	to_chat(c, "<span class='notice'>***********************************************************\n\
		Left Mouse Button on anything to delete it. If you break it, you buy it.\n\
		Right Mouse Button on anything to delete everything of the type. Probably don\'t do this unless you know what you are doing.\n\
		***********************************************************</span>")
```

File: code/modules/buildmode/submodes/outfit.dm [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+/datum/buildmode_mode/outfit/Destroy()
+	dressuptime = null
+	return ..()
+
+/datum/buildmode_mode/outfit/show_help(client/c)
+	to_chat(c, "<span class='notice'>***********************************************************</span>")
+	to_chat(c, "<span class='notice'>Right Mouse Button on buildmode button = Select outfit to equip.</span>")
+	to_chat(c, "<span class='notice'>Left Mouse Button on mob/living/carbon/human = Equip the selected outfit.</span>")
+	to_chat(c, "<span class='notice'>Right Mouse Button on mob/living/carbon/human = Strip and delete current outfit.</span>")
+	to_chat(c, "<span class='notice'>***********************************************************</span>")
```
Comment (Rohesie) [inline code/modules/buildmode/submodes/outfit.dm:14]:
```suggestion
	to_chat(c, "<span class='notice'>***********************************************************\n\
		Right Mouse Button on buildmode button = Select outfit to equip.\n\
		Left Mouse Button on mob/living/carbon/human = Equip the selected outfit.\n\
		Right Mouse Button on mob/living/carbon/human = Strip and delete current outfit.\n\
		***********************************************************</span>")
```

**Pull request merged**